multitenant: handle missing NodeDescriptor in crdb_internal.ranges_no_leases

ecwall opened the pull request:
Fixes #92915

This change matches the previous behavior of using "" for locality if the NodeDescriptor is not found instead of returning an error when generating crdb_internal.ranges_no_leases.

Release note: None

ecwall committed:
multitenant: handle missing NodeDescriptor in crdb_internal.ranges_no_leases

Fixes #92915

This change matches the previous behavior of using "" for locality if the
NodeDescriptor is not found instead of returning an error when generating
crdb_internal.ranges_no_leases.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/93127)
<!-- Reviewable:end -->

knz reviewed (commented):
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93127)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @rafiss)

___
*[`pkg/sql/crdb_internal.go` line 3728 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-NIc0DOO8_wnBmNMgDwF:-NIc0DOO8_wnBmNMgDwG:bx14fl9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a067b08530198dc4c76896783bc68fa4b829c8b1/pkg/sql/crdb_internal.go#L3728)):*
> ```Go
> 				// Use empty string if nodeDesc is not found.
> 				// See https://github.com/cockroachdb/cockroach/issues/92915.
> 				if err == nil {
> ```

Not all errors indicate the desc was not found. Some errors should be propagated. Please use a more precise predicate here. Something like:

```go
if err != nil {
   if !xxx(err) {
       return err
   }
} else {
  replicaLocalityString = nodeDesc.Locality.String()
}
```
 


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Nice, thanks for the quick fix! To confirm, the linked roachtest passes with this patch, right?

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-:-NIc1DtD3gD_ezPcUGTG:b-fyrdxy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[`pkg/sql/crdb_internal.go` line 3726 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-NIc0VCC-tk73bx5WeoC:-NIc0VCC-tk73bx5WeoD:b-61ikzi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a067b08530198dc4c76896783bc68fa4b829c8b1/pkg/sql/crdb_internal.go#L3726)):*
> ```Go
> 				var replicaLocalityString string
> 				nodeDesc, err := p.ExecCfg().NodeDescs.GetNodeDescriptor(replica.NodeID)
> 				// Use empty string if nodeDesc is not found.
> ```

nit: could you expand on this comment a bit to say that we don't want this table to not render if locality information isn't available for a particular node for some reason?


<!-- Sent from Reviewable.io -->

ecwall commented:
@arulajmani @knz getting a "Sorry, you've run out of GitHub API quota." error in reviewable, but I addressed your comments.

Working on running `roachtest` now.

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93127)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, and @rafiss)

___
*[`pkg/sql/crdb_internal.go` line 3726 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-NIc0VCC-tk73bx5WeoC:-NIc6FrCADlpsT2PN-xW:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/a067b08530198dc4c76896783bc68fa4b829c8b1/pkg/sql/crdb_internal.go#L3726)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: could you expand on this comment a bit to say that we don't want this table to not render if locality information isn't available for a particular node for some reason?
</blockquote></details>

Updated.
___
*[`pkg/sql/crdb_internal.go` line 3728 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-NIc0DOO8_wnBmNMgDwF:-NIc5BP-EU-Pil9I1vvK:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/a067b08530198dc4c76896783bc68fa4b829c8b1/pkg/sql/crdb_internal.go#L3728)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Not all errors indicate the desc was not found. Some errors should be propagated. Please use a more precise predicate here. Something like:

```go
if err != nil {
   if !xxx(err) {
       return err
   }
} else {
  replicaLocalityString = nodeDesc.Locality.String()
}
```
 
</blockquote></details>

Updated.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 10 of 13 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93127)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @rafiss)

___
*[`pkg/util/errorutil/descriptor.go` line 20 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-NIcBFis2s2AzgagKsoK:-NIcBFis2s2AzgagKsoL:bawowh0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d382d4c3a5cec1cce87ca6d586afe5c4d86e27ce/pkg/util/errorutil/descriptor.go#L20)):*
> ```Go
> )
> 
> const descriptorNotFound = "unable to look up descriptor for"
> ```

The API here is fine, but the implementation is not. We don't use `Error()` comparisons any more in CockroachDB.

Try this:

```go
type descriptorNotFound struct{msg string}
func (e *descriptorNotFound) Error() string { return e.msg }
```
___
*[`pkg/util/errorutil/descriptor.go` line 23 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-NIcBdWg6p5t8Z2k7f17:-NIcBdWg6p5t8Z2k7f18:b98t3dx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d382d4c3a5cec1cce87ca6d586afe5c4d86e27ce/pkg/util/errorutil/descriptor.go#L23)):*
> ```Go
> 
> func IsDescriptorNotFoundError(err error) bool {
> 	return strings.HasPrefix(err.Error(), descriptorNotFound)
> ```

return `errors.HasType(err, (*descriptorNotFound)(nil))`
___
*[`pkg/util/errorutil/descriptor.go` line 27 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-NIcBhzlERqknpe087Cw:-NIcBhzlERqknpe087Cx:b-c2hskd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d382d4c3a5cec1cce87ca6d586afe5c4d86e27ce/pkg/util/errorutil/descriptor.go#L27)):*
> ```Go
> 
> func NewNodeNotFoundError(nodeID roachpb.NodeID) error {
> 	return errors.Errorf("%s n%d", descriptorNotFound, nodeID)
> ```

`return &descriptorNotFound{fmt.Sprintf("unable to look up descriptor for n%d", nodeID")}`
___
*[`pkg/util/errorutil/descriptor.go` line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-NIcBt0N2rc5JEe1vc9c:-NIcBt0N2rc5JEe1vc9d:b-ltj36v) ([raw file](https://github.com/cockroachdb/cockroach/blob/d382d4c3a5cec1cce87ca6d586afe5c4d86e27ce/pkg/util/errorutil/descriptor.go#L31)):*
> ```Go
> 
> func NewStoreNotFoundError(storeID roachpb.StoreID) error {
> 	return errors.Errorf("%s storeID %d", descriptorNotFound, storeID)
> ```

ditto & reuse the original error message for stores


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93127)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @rafiss)

___
*[`pkg/sql/crdb_internal.go` line 3726 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-NIc0VCC-tk73bx5WeoC:-NIcDXJw3kQA0d6iKolH:b32bhhg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a067b08530198dc4c76896783bc68fa4b829c8b1/pkg/sql/crdb_internal.go#L3726)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Updated.
</blockquote></details>

Arul: what do you think, should we render NULL in that case?


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93127)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, and @rafiss)

___
*[`pkg/util/errorutil/descriptor.go` line 20 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-NIcBFis2s2AzgagKsoK:-NIcH1ZQ6zYJFIU6pBE9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d382d4c3a5cec1cce87ca6d586afe5c4d86e27ce/pkg/util/errorutil/descriptor.go#L20)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

The API here is fine, but the implementation is not. We don't use `Error()` comparisons any more in CockroachDB.

Try this:

```go
type descriptorNotFound struct{msg string}
func (e *descriptorNotFound) Error() string { return e.msg }
```

</blockquote></details>

Done.
___
*[`pkg/util/errorutil/descriptor.go` line 23 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-NIcBdWg6p5t8Z2k7f17:-NIcH4s9BZ64DW8nwurK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d382d4c3a5cec1cce87ca6d586afe5c4d86e27ce/pkg/util/errorutil/descriptor.go#L23)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

return `errors.HasType(err, (*descriptorNotFound)(nil))`
</blockquote></details>

Done.
___
*[`pkg/util/errorutil/descriptor.go` line 27 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-NIcBhzlERqknpe087Cw:-NIcH5VwCWF1Oyd9nDTm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d382d4c3a5cec1cce87ca6d586afe5c4d86e27ce/pkg/util/errorutil/descriptor.go#L27)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

`return &descriptorNotFound{fmt.Sprintf("unable to look up descriptor for n%d", nodeID")}`
</blockquote></details>

Done.
___
*[`pkg/util/errorutil/descriptor.go` line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-NIcBt0N2rc5JEe1vc9c:-NIcH647AoHlq_FOJT5u:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d382d4c3a5cec1cce87ca6d586afe5c4d86e27ce/pkg/util/errorutil/descriptor.go#L31)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

ditto & reuse the original error message for stores
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
LGTM but let's let arul finalize the review

> Reviewed 2 of 13 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-:-NIcJvDtALbtIn_J_Ng_:b4diiwl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @rafiss)



<!-- Sent from Reviewable.io -->

ecwall commented:
I'm getting a different error trying to reproduce this:
```
test_impl.go:347: test failure #1: (test_impl.go:291).Fatal: output in run_183835.339062000_n1-4_workload_init_schemachange: ./workload init schemachange returned: COMMAND_PROBLEM: exit status 126
```

arulajmani reviewed (approved):
LGTM once we've confirmed the roachtest passes. 

> Reviewed 10 of 13 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-:-NIcptUWEcOK6FDfgsDH:b-h7ivqi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @knz, and @rafiss)

___
*[`pkg/sql/crdb_internal.go` line 3726 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-NIc0VCC-tk73bx5WeoC:-NIchZ0k7ccVbyyyldMF:bmpmvo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a067b08530198dc4c76896783bc68fa4b829c8b1/pkg/sql/crdb_internal.go#L3726)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Arul: what do you think, should we render NULL in that case?
</blockquote></details>

I think it's a good idea -- that way, we aren't conflating no locality string with one that couldn't be found. 


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93127)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, and @rafiss)

___
*[`pkg/sql/crdb_internal.go` line 3726 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93127#-NIc0VCC-tk73bx5WeoC:-NInBbS50PO3Ic9HMPG3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a067b08530198dc4c76896783bc68fa4b829c8b1/pkg/sql/crdb_internal.go#L3726)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I think it's a good idea -- that way, we aren't conflating no locality string with one that couldn't be found. 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ecwall commented:
bors r=arulajmani
